File: src/backtrack/39.combination-sum.ts
/*
 * @lc app=leetcode.cn id=39 lang=typescript
 *
 * [39] 组合总和
 */

// @lc code=start
function combinationSum(candidates: number[], target: number): number[][] {
  let res = [];

  // 记录路径和路径和
  let path = [];
  let pathSum = 0;

  backtrack(0);

  return res;

  function backtrack(start) {

    if (pathSum === target) {
      res.push(path.slice());
      return;
    }

    for (let [i, v] of candidates.entries()) {

      // 组合问题，不可以重复使用同一个元素
      if (i < start) continue;

      // 路径和即将大于 target，跳过
      if (pathSum + v > target) continue;

      pathSum += v;
      path.push(v);
      backtrack(i);
      path.pop();
      pathSum -= v;
    }
  }
}
// @lc code=end

export {}

File: src/bin/39.combination-sum.py
#
# @lc app=leetcode.cn id=39 lang=python3
# @lcpr version=30204
#
# [39] 组合总和
#


# @lcpr-template-start
from typing import List, Optional
from src.adt.py.leetcodeType import ListNode, TreeNode
# @lcpr-template-end
# @lc code=start
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        
        path = []
        pathSum = 0
        res = []
        n = len(nums)

        def comb(start):
            nonlocal pathSum

            if pathSum > target:
                return
            
            if pathSum == target:
                res.append(path.copy())
                return

            for i in range(start, n):
                path.append(nums[i])
                pathSum += nums[i]
                comb(i)
                pathSum -= nums[i]
                path.pop()

        comb(0)

        return res


# @lc code=end



#
# @lcpr case=start
# [2,3,6,7]\n7\n
# @lcpr case=end

# @lcpr case=start
# [2,3,5]\n8\n
# @lcpr case=end

# @lcpr case=start
# [2]\n1\n
# @lcpr case=end

#

